pkg/controller/backup_storage_location_controller.go

```diff
@@ -0,0 +1,210 @@
+		return
+	}
+
+	pluginManager := c.newPluginManager(c.logger)
+	defer pluginManager.CleanupClients()
+
+	var unavailable []string
+	log := c.logger
+	for _, location := range locations {
+		locationName := location.Name
+		log = c.logger.WithField("backupLocation", locationName)
+
+		storeValidationPeriod := c.defaultStoreValidationPeriod
+		if location.Spec.StoreValidationPeriod != nil {
+			storeValidationPeriod = location.Spec.StoreValidationPeriod.Duration
+			if storeValidationPeriod == 0 {
+				log.Debug("Validation period for this location is set to 0, skipping validation")
+				continue
+			}
+
+			if storeValidationPeriod < 0 {
+				log.Debug("Validation period must be non-negative")
+				storeValidationPeriod = c.defaultStoreValidationPeriod
+			}
+		}
+
+		lastValidation := location.Status.LastValidationTime
+		if lastValidation != nil {
+			log.Debug("Checking if location needs to be validated at this time")
+			nextValidation := lastValidation.Add(storeValidationPeriod)
+			if time.Now().UTC().Before(nextValidation) {
+				continue
+			}
+		}
+
+		log.Debug("Location is ready to be validated")
+
+		backupStore, err := c.newBackupStore(location, pluginManager, log)
+		if err != nil {
+			unavailable = append(unavailable, errors.Wrapf(err, "error getting backup store for location %q", locationName).Error())
+			continue
+		}
+
+		if err := backupStore.IsValid(); err != nil {
+			// update status
+			err2 := c.patchBackupStorageLocation(location, func(r *velerov1api.BackupStorageLocation) {
+				r.Status.Phase = velerov1api.BackupStorageLocationPhaseUnavailable
+				r.Status.LastValidationTime = &metav1.Time{Time: time.Now().UTC()}
+			})
+			if err2 != nil {
+				unavailable = append(unavailable, errors.Wrapf(err2, "error updating location %q to %s status", locationName, velerov1api.BackupStorageLocationPhaseUnavailable).Error())
+			} else {
+				unavailable = append(unavailable, errors.Wrapf(err, "location %q is unavailable", locationName).Error())
+			}
+			log.Debug("Location has been updated to Unavailable")
+		} else {
+			// update status
+			err := c.patchBackupStorageLocation(location, func(r *velerov1api.BackupStorageLocation) {
+				r.Status.Phase = velerov1api.BackupStorageLocationPhaseAvailable
+				r.Status.LastValidationTime = &metav1.Time{Time: time.Now().UTC()}
+			})
+			if err != nil {
+				unavailable = append(unavailable, errors.Wrapf(err, "error updating location %q to %s status", locationName, velerov1api.BackupStorageLocationPhaseAvailable).Error())
+			}
+			log.Debug("Location has been updated to Available")
+		}
+	}
+
+	var defaultFound bool
+	if len(unavailable) == len(locations) { // no BSL available
+		log.Errorf("There are no locations available, at least one available location is required: %s", strings.Join(unavailable, "; "))
+	} else if len(unavailable) > 0 { // some but not all BSL unavailable
+		for _, location := range locations {
```

Comment:
Seems like we are looping through BSLs just to log if the `default` BSL is unavailable.
Can we move L#170-l#174 inside the loop at L105?

Reply:
Yes, I cleaned this up, thanks.